src/jv_aux.c

```diff
@@ -500,8 +500,10 @@ jv jv_delpaths(jv object, jv paths) {
     jv_free(paths);
     return jv_invalid_with_msg(jv_string("Paths must be specified as an array"));
   }
-  paths = jv_sort(paths, jv_copy(paths));
-  jv_array_foreach(paths, i, elem) {
+  int paths_length = jv_array_length(jv_copy(paths));
+  paths = jv_sort(jv_array_slice(paths, 0, paths_length), jv_copy(paths));
```

Comment:
This can't be the fix, right?  I mean, the problem is that we need to canonicalize array indices in `paths` first, no?

Reply:
`jv_array_slice(a, 0, length_of_a)` should be a no-op, too.

Reply:
I don't understand what this patch is trying to do either

Reply:
@nicowilliams 
is this what you're suggesting?
```
jv canonicalize_indices(jv paths) {
  int paths_length = jv_array_length(jv_copy(paths));
  for (int i = 0; i < paths_length; i++) {
    jv path = jv_array_get(jv_copy(paths), i);
    int index = jv_number_value(path);
    if (index < 0) {
      index = paths_length + index;
    }
    if (index < 0 || index >= paths_length) {
      jv_free(paths);
      return jv_invalid_with_msg(jv_string("Invalid array index"));
    }
    paths = jv_array_set(paths, i, jv_number(index));
  }
  return paths;
}
```

Reply:
To canonicalize the paths we need to refer to the tree they refer to.  The idea is to take any negative indices in the paths and normalize them to positive indices, but to do this you need to know the length of the corresponding array in the tree for that negative index in the path.

Reply:
`if (index < 0) index = paths_length + index;`? Why would you want to do that? What bug did you solve with those patches?

Reply:
Okay so to convert a negative index into a positive index, we need the length of the subtree, this could work then
```
jv canonicalize_indices(jv paths, jv tree) {
  int paths_length = jv_array_length(jv_copy(paths));
  for (int i = 0; i < paths_length; i++) {
    jv path = jv_array_get(jv_copy(paths), i);
    jv subtree = tree;
    int path_length = jv_array_length(jv_copy(path));
    for (int j = 0; j < path_length; j++) {
      jv step = jv_array_get(jv_copy(path), j);
      if (jv_is_number(step)) {
        int index = jv_number_value(step);
        if (index < 0) {
          int subtree_length = jv_array_length(jv_copy(subtree));
          index = subtree_length + index;
          if (index < 0) {
            // Invalid index, handle error
            jv_free(paths);
            return jv_invalid_with_msg(jv_string("Invalid array index"));
          }
          path = jv_array_set(path, j, jv_number(index));
        }
        subtree = jv_array_get(jv_copy(subtree), index);
      } else if (jv_is_string(step)) {
        subtree = jv_object_get(jv_copy(subtree), step);
      } else {
        // Invalid path step, handle error
        jv_free(paths);
        return jv_invalid_with_msg(jv_string("Invalid path step"));
      }
    }
    paths = jv_array_set(paths, i, path);
  }
  return paths;
}
```

Reply:
With canonicalising a path, we mean like converting:
```json
["foo",{"start":10,"end":null},{"start":-200,"end":-2},20]
["foo",-20,"hello"]
```
To (if foo is an array with 120 elements)
```json
["foo",30]
["foo",100,"hello"]
```
You need to know the input to do that; if foo were an array with 300 elements you would want this for example:
```json
["foo",120]
["foo",280]
```

Reply:
Yeah, it's a bit messy.  Because path elements could overlap because of the slice specifiers, we might have to explode `{"start":...,"end":...}` into the individual indices that match those slice specifiers.

@capak07 the context here is that internally a "slice" specifier like `$something[start:end]` is represented as an object with a `"start"` and `"end"` keys, and that the start and end indices can be negative to mean "starting from the end of this string or array".  Here `jv_sort()` just can't possibly do the right thing.  We should either have a `jv_sort_paths()` that does, or we should normalize the `paths` to be given to `jv_sort()`, and the latter seems like the better path forward for now because the former will likely require that we use `qsort_r()` (which isn't portable, so we might have to import one from a BSD).

Reply:
You should not reject object indices (they are slices), and `del(.[100])` on a short array should not be an error, so you should not start throwing an invalid path error